cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataScanner.java

```diff
@@ -0,0 +1,236 @@
+    final List<DeleteEntry> toDelete = new ArrayList<>();
+    if (transactionTimeout < CUTOFF_DISCOUNT) {
+      LOG.warn("Transaction timeout for log cleanup is really low {}s", transactionTimeout);
+    }
+    final int cutOffTransactionTime =
+      transactionTimeout > CUTOFF_DISCOUNT ? transactionTimeout - CUTOFF_DISCOUNT : transactionTimeout;
+    try {
+      transactional.execute(transactionTimeout, new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          Stopwatch stopwatch = new Stopwatch();
+          stopwatch.start();
+
+          Table table = LoggingStoreTableUtil.getMetadataTable(context, datasetManager);
+          byte[] startRowKey = NEW_ROW_KEY_PREFIX;
+          byte[] endRowKey = NEW_ROW_KEY_PREFIX_END;
+          Scanner scanner;
+          boolean reachedEnd = false;
+          while (!reachedEnd) {
+            if (stopwatch.elapsedTime(TimeUnit.SECONDS) > cutOffTransactionTime) {
+              break;
+            }
+            scanner = table.scan(startRowKey, endRowKey);
+            Row row;
+            while ((row = scanner.next()) != null) {
+              byte[] rowkey = row.getRow();
+              // file creation time is the last 8-bytes in rowkey in the new format
+              long creationTime = Bytes.toLong(rowkey, rowkey.length - Bytes.SIZEOF_LONG, Bytes.SIZEOF_LONG);
+              if (creationTime <= tillTime) {
+                // expired - can be deleted
+                toDelete.add(
+                  new DeleteEntry(rowkey,
+                                  new URI(Bytes.toString(row.get(LoggingStoreTableUtil.META_TABLE_COLUMN_KEY)))));
+              } else {
+                // update start-row key based on the logging context and start a new scan.
+                startRowKey = Bytes.add(NEW_ROW_KEY_PREFIX, getNextContextStartKey(rowkey));
+                scanner.close();
+                break;
+              }
+            }
+            // if row is null, then scanner next returned null. so we have reached the end.
+            if (row == null) {
+              reachedEnd = true;
+              scanner.close();
+            }
+          }
+        }
+      });
+    } catch (TransactionFailureException e) {
+      LOG.warn("Got Exception while scanning metadata table", e);
+      // if there is an exception, no metadata, so delete file should be skipped.
+      return new ArrayList<>();
+    }
+
+    if (!toDelete.isEmpty()) {
+      // we will call delete on old metadata even whenever there is expired entries to delete in new format.
+      // though the first call will delete all old meta data.
+      scanAndDeleteOldMetaData(transactionTimeout, cutOffTransactionTime);
+
+      // delete meta data first
+      return deleteNewMetadataEntries(toDelete, transactionTimeout, cutOffTransactionTime);
+    }
+    // toDelete is empty, safe to return that
+    return toDelete;
+  }
+
+
+  class DeleteEntry {
```

Comment:
Define inner class after all methods. Also, this class should be `static final`.